In [1]:
import numpy as np
import pandas as pd
import datetime as dt

In [2]:
today = pd.Timestamp('2024-12-31')

In [5]:
df = pd.DataFrame({
    'customer_id': [1, 2, 3, 4, 5],
    'last_purchase_date': [
        '2024-12-01',
        '2024-09-15',
        '2024-01-10',
        '2024-11-20',
        '2023-06-30'
    ],
    'contract_expiry_date': [
        '2025-01-15',
        '2025-06-30',
        '2024-12-15',
        '2025-03-01',
        '2025-02-28']})

df['last_purchase_date'] = pd.to_datetime(df["last_purchase_date"])
df["contract_expiry_date"] = pd.to_datetime(df["contract_expiry_date"])

df

,customer_id,last_purchase_date,contract_expiry_date
0,1,2024-12-01,2025-01-15
1,2,2024-09-15,2025-06-30
2,3,2024-01-10,2024-12-15
3,4,2024-11-20,2025-03-01
4,5,2023-06-30,2025-02-28


### Create Time Since Features

In [6]:
# days since last purchase
df['days_since_last_purchase'] = (today - df['last_purchase_date']).dt.days
df

,customer_id,last_purchase_date,contract_expiry_date,days_since_last_purchase
0,1,2024-12-01,2025-01-15,30
1,2,2024-09-15,2025-06-30,107
2,3,2024-01-10,2024-12-15,356
3,4,2024-11-20,2025-03-01,41
4,5,2023-06-30,2025-02-28,550


### Create Time Until Features

In [7]:
# days until contract expiry
df["days_until_contract_expiry"] = (df["contract_expiry_date"] - today).dt.days
df

,customer_id,last_purchase_date,contract_expiry_date,days_since_last_purchase,days_until_contract_expiry
0,1,2024-12-01,2025-01-15,30,15
1,2,2024-09-15,2025-06-30,107,181
2,3,2024-01-10,2024-12-15,356,-16
3,4,2024-11-20,2025-03-01,41,60
4,5,2023-06-30,2025-02-28,550,59


## Questions

##### 1- Which customer has been inactive the longest? How many days since their last purchase?
##### Ans: Customer 5. 550 days ago they made thier last purchase

##### 2- Which customer's contract is expiring soonest? How many days do they have left?
##### Ans: Customer 1. Only 15 days left

##### 3- Customer 3's contract_expiry_date is 2024-12-15 — Which is already in the past relative to today (2024-12-31). What value does days_until_contract_expiry give for them? Is that a problem? What would you do about it?
##### Ans: It gives -16. Since not many days have gone by, we can give them few reminders and wait for few days before terminating the contract
